README.md
# About this stack

A ROS1 stack that handles communication between Rover Robotics rover platforms and computer payloads.
The rover platform accept velocity commands and publishes wheel encoder data and power information.

Documentation for this package can be found at http://wiki.ros.org/rr_openrover_stack

Master|Develop|Kinetic-Devel
---|---|---
[![CircleCI](https://circleci.com/gh/RoverRobotics/rr_openrover_stack/tree/master.svg?style=svg)](https://circleci.com/gh/RoverRobotics/rr_openrover_stack/tree/master)|[![CircleCI](https://circleci.com/gh/RoverRobotics/rr_openrover_stack/tree/develop.svg?style=svg)](https://circleci.com/gh/RoverRobotics/rr_openrover_stack/tree/develop)|[![CircleCI](https://circleci.com/gh/RoverRobotics/rr_openrover_stack/tree/kinetic-devel.svg?style=svg)](https://circleci.com/gh/RoverRobotics/rr_openrover_stack/tree/kinetic-devel)

# Robot Computer Installation Guide

This guide is to help you setup your Computer with Rover Packages to able to run Rover Robotics Robots

## Prerequisites / Requirements
1. Ubuntu 18.0.4
2. Git
3. Nano
4. ROS Melodic
5. Rosdep
## Installation
Open a Terminal and Navigate to your ROS Workspace

Install missing ros packages with
```bash
sudo apt install ros-melodic-joy ros-melodic-serial ros-melodic-twist-mux ros-melodic-tf2-geometry-msgs ros-melodic-robot-localization ros-melodic-gmapping ros-melodic-move-base -y
```
Go to src folder
```bash
cd src/
```
Use the git to clone our repository

```bash
git clone https://github.com/roverrobotics/rr_openrover_stack rr_openrover_stack
```
Clone ps4 ROS driver(optional). Need Driver from Below if install.
```bash
git clone https://github.com/naoki-mizuno/ds4_driver
```

In order to consistently detect the robots every reboot, we need to setup the Udev rules for it. Default UDEV rules are in our our stack. You can copy it via
```bash
sudo cp src/rr_openrover_stack/udev/roverrobotics.rules /etc/udev/rules.d/
sudo udevadm control --reload-rules && sudo udevadm trigger
```
## Installing Dual Shock 4 (Ps4 Controller Driver)
The following steps are to help you 
install DualShock 4 (Ps4 Controller Driver) that being use as a default method for our robots. You may skip this step if you are using another controller.

Make a new directory to store the driver.
```bash
mkdir ~/drivers
cd ~/drivers
```
Clone and Install Ds4driver
```bash
git clone https://github.com/naoki-mizuno/ds4drv --branch devel
mkdir -p ~/.local/lib/python2.7/site-packages
python2 setup.py install --prefix ~/.local
sudo cp udev/50-ds4drv.rules /etc/udev/rules.d/
```
Now you should be able to build your ROS Workspace

Navigate to ROS workspace. 

Run this to update your dependencies.
```bash
rosdep install --from-paths src --ignore-src -r -y
```
Build and source your workspace
```bash
catkin_make
source /devel/setup.bash
```

## Usage

Your Robot can now be control via our Launch Files. *** These are template launch files. You may need to add or remove nodes if your robot are configured with a different hardware***

Rover Pro:
```bash
roslaunch rr_openrover_driver teleop_launch
```
Rover Pro with slam pack:
```bash
roslaunch rr_openrover_driver slampack_launch
```

Rover Zero:
```bash
roslaunch rr_roverzero_driver teleop_launch
```
Rover Zero V2:
```bash
roslaunch vesc_driver vesc_driver_node.launch
```
# Other Installation

## Systemd startup scripts
We have prepared some system startup scripts that would automaticly launch the robot to be able to control with the ps4 controller on power up. 

This is useful for those who just want to drive the robot around.

Run this command to inside your ROS Workspace install the startup script
```bash
sudo src/rr_openrover_stack/opscripts/install.sh


rr_openrover_stack/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package rr_openrover_stack
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

<<<<<<< HEAD
1.0.1 (2020-06-15)
------------------
=======
1.1.1 (2020-06-15)
------------------
* add pyserial to package.xml
* Contributors: padiln

1.1.0 (2020-06-08)
------------------
* release to melodic
>>>>>>> upstream/melodic_with_v2support

1.0.0 (2020-06-08)
------------------
* Merge pull request `#76 <https://github.com/RoverRobotics/rr_openrover_stack/issues/76>`_ from RoverRobotics/feature/padiln/expose_odometry
  Feature/padiln/expose odometry
* Merge branch 'develop' into feature/padiln/rviz_gazebo_sim
* Merge pull request `#51 <https://github.com/RoverRobotics/rr_openrover_stack/issues/51>`_ from RoverRobotics/master
  Sync with master
* Merge pull request `#43 <https://github.com/RoverRobotics/rr_openrover_stack/issues/43>`_ from RoverRobotics/kinetic-devel
  Merging change log update from bloom release
* Merge pull request `#40 <https://github.com/RoverRobotics/rr_openrover_stack/issues/40>`_ from RoverRobotics/kinetic-devel
  Kinetic devel
* Contributors: padiln

0.8.0 (2020-03-25)
------------------

0.7.4 (2020-02-19)
------------------
* Changed maintainership of all packages
* Fixed PID overwrite issue

0.7.3 (2019-10-14)
------------------
* Migrating all usage of tf to tf2
* Simplified rover data messages and only publish rate-based data in legacy mode

0.7.2 (2019-07-18)
------------------
* Collect the packages to operate the rover platform into a stack
* Changed repo name from rr_openrover_ros1 to rr_openrover_stack
* Refactor rr_control_input_manager from a MUX to a TwistStamped latency manager

0.7.1 (2019-06-24)
------------------

0.7.0 (2019-05-13)
------------------

0.6.2 (2019-01-23)
------------------

0.6.1 (2018-12-20)
------------------

0.6.0 (2018-12-07)
------------------

0.5.1 (2018-09-20)
------------------

0.5.0 (2018-09-06)
------------------

0.4.0 (2018-07-23)
------------------

0.3.0 (2018-07-19)
------------------

0.2.0 (2018-07-17)
------------------

0.1.0 (2018-07-10)
------------------


vesc_driver/src/vesc_packet.cpp
// -*- mode:c++; fill-column: 100; -*-

#include "vesc_driver/vesc_packet.h"

#include <cassert>
#include <iterator>

#include <boost/range/begin.hpp>
#include <boost/range/distance.hpp>
#include <boost/range/end.hpp>

#include "vesc_driver/vesc_packet_factory.h"
#include "vesc_driver/datatypes.h"

namespace vesc_driver
{

VescFrame::VescFrame(int payload_size)
{
  assert(payload_size >= 0 && payload_size <= 1024);

  if (payload_size < 256) {
    // single byte payload size
    frame_.reset(new Buffer(VESC_MIN_FRAME_SIZE + payload_size));
    *frame_->begin() = 2;
    *(frame_->begin() + 1) = payload_size;
    payload_.first = frame_->begin() + 2;
  }
  else {
    // two byte payload size
    frame_.reset(new Buffer(VESC_MIN_FRAME_SIZE + 1 + payload_size));
    *frame_->begin() = 3;
    *(frame_->begin() + 1) = payload_size >> 8;
    *(frame_->begin() + 2) = payload_size & 0xFF;
    payload_.first = frame_->begin() + 3;
  }

  payload_.second = payload_.first + payload_size;
  *(frame_->end() - 1) = 3;
}

VescFrame::VescFrame(const BufferRangeConst& frame, const BufferRangeConst& payload)
{
  /* VescPacketFactory::createPacket() should make sure that the input is valid, but run a few cheap
     checks anyway */
  assert(boost::distance(frame) >= VESC_MIN_FRAME_SIZE);
  assert(boost::distance(frame) <= VESC_MAX_FRAME_SIZE);
  assert(boost::distance(payload) <= VESC_MAX_PAYLOAD_SIZE);
  assert(std::distance(frame.first, payload.first) > 0 &&
         std::distance(payload.second, frame.second) > 0);

  frame_.reset(new Buffer(boost::begin(frame), boost::end(frame)));
  payload_.first = frame_->begin() + std::distance(frame.first, payload.first);
  payload_.second = frame_->begin() + std::distance(frame.first, payload.second);
}

VescPacket::VescPacket(const std::string& name, int payload_size, int payload_id) :
  VescFrame(payload_size), name_(name)
{
  assert(payload_id >= 0 && payload_id < 256);
  assert(boost::distance(payload_) > 0);
  *payload_.first = payload_id;
}

VescPacket::VescPacket(const std::string& name, boost::shared_ptr<VescFrame> raw) :
  VescFrame(*raw), name_(name)
{
}

/*------------------------------------------------------------------------------------------------*/

VescPacketFWVersion::VescPacketFWVersion(boost::shared_ptr<VescFrame> raw) :
  VescPacket("FWVersion", raw)
{
}

int VescPacketFWVersion::fwMajor() const
{
  return *(payload_.first + 1);
}

int VescPacketFWVersion::fwMinor() const
{
  return *(payload_.first + 2);
}

REGISTER_PACKET_TYPE(COMM_FW_VERSION, VescPacketFWVersion)

VescPacketRequestFWVersion::VescPacketRequestFWVersion() :
  VescPacket("RequestFWVersion", 1, COMM_FW_VERSION)
{
  VescFrame::CRC crc_calc;
  crc_calc.process_bytes(&(*payload_.first), boost::distance(payload_));
  uint16_t crc = crc_calc.checksum();
  *(frame_->end() - 3) = static_cast<uint8_t>(crc >> 8);
  *(frame_->end() - 2) = static_cast<uint8_t>(crc & 0xFF);
}

/*------------------------------------------------------------------------------------------------*/

VescPacketValues::VescPacketValues(boost::shared_ptr<VescFrame> raw) :
  VescPacket("Values", raw)
{
}

double VescPacketValues::temp_mos1() const
{
  int16_t v = static_cast<int16_t>((static_cast<uint16_t>(*(payload_.first + 1)) << 8) +
                                   static_cast<uint16_t>(*(payload_.first + 2)));
  return static_cast<double>(v) / 10.0;
}
double VescPacketValues::temp_mos2() const
{
  int16_t v = static_cast<int16_t>((static_cast<uint16_t>(*(payload_.first + 3)) << 8) +
                                   static_cast<uint16_t>(*(payload_.first + 4)));
  return static_cast<double>(v) / 10.0;
}
double VescPacketValues::temp_mos3() const
{
  int16_t v = static_cast<int16_t>((static_cast<uint16_t>(*(payload_.first + 5)) << 8) +
                                   static_cast<uint16_t>(*(payload_.first + 6)));
  return static_cast<double>(v) / 10.0;
}
double VescPacketValues::temp_mos4() const
{
  int16_t v = static_cast<int16_t>((static_cast<uint16_t>(*(payload_.first + 7)) << 8) +
                                   static_cast<uint16_t>(*(payload_.first + 8)));
  return static_cast<double>(v) / 10.0;
}
double VescPacketValues::temp_mos5() const
{
  int16_t v = static_cast<int16_t>((static_cast<uint16_t>(*(payload_.first + 9)) << 8) +
                                   static_cast<uint16_t>(*(payload_.first + 10)));
  return static_cast<double>(v) / 10.0;
}
double VescPacketValues::temp_mos6() const
{
  int16_t v = static_cast<int16_t>((static_cast<uint16_t>(*(payload_.first + 11)) << 8) +
                                   static_cast<uint16_t>(*(payload_.first + 12)));
  return static_cast<double>(v) / 10.0;
}
double VescPacketValues::temp_pcb() const
{
  int16_t v = static_cast<int16_t>((static_cast<uint16_t>(*(payload_.first + 13)) << 8) +
                                   static_cast<uint16_t>(*(payload_.first + 14)));
  return static_cast<double>(v) / 10.0;
}
double VescPacketValues::current_motor() const
{
  int32_t v = static_cast<int32_t>((static_cast<uint32_t>(*(payload_.first + 15)) << 24) +
                                   (static_cast<uint32_t>(*(payload_.first + 16)) << 16) +
                                   (static_cast<uint32_t>(*(payload_.first + 17)) << 8) +
                                   static_cast<uint32_t>(*(payload_.first + 18)));
  return static_cast<double>(v) / 100.0;
}
double VescPacketValues::current_in() const
{
  int32_t v = static_cast<int32_t>((static_cast<uint32_t>(*(payload_.first + 19)) << 24) +
                                   (static_cast<uint32_t>(*(payload_.first + 20)) << 16) +
                                   (static_cast<uint32_t>(*(payload_.first + 21)) << 8) +
                                   static_cast<uint32_t>(*(payload_.first + 22)));
  return static_cast<double>(v) / 100.0;
}
double VescPacketValues::duty_now() const
{
  int16_t v = static_cast<int16_t>((static_cast<uint16_t>(*(payload_.first + 23)) << 8) +
                                   static_cast<uint16_t>(*(payload_.first + 24)));
  return static_cast<double>(v) / 1000.0;
}
double VescPacketValues::rpm() const
{
  int32_t v = static_cast<int32_t>((static_cast<uint32_t>(*(payload_.first + 25)) << 24) +
                                   (static_cast<uint32_t>(*(payload_.first + 26)) << 16) +
                                   (static_cast<uint32_t>(*(payload_.first + 27)) << 8) +
                                   static_cast<uint32_t>(*(payload_.first + 28)));
  return static_cast<double>(v);
}
double VescPacketValues::v_in() const
{
  int16_t v = static_cast<int16_t>((static_cast<uint16_t>(*(payload_.first + 29)) << 8) +
                                   static_cast<uint16_t>(*(payload_.first + 30)));
  return static_cast<double>(v) / 10.0;
}
double VescPacketValues::amp_hours() const
{
  int32_t v = static_cast<int32_t>((static_cast<uint32_t>(*(payload_.first + 31)) << 24) +
                                   (static_cast<uint32_t>(*(payload_.first + 32)) << 16) +
                                   (static_cast<uint32_t>(*(payload_.first + 33)) << 8) +
                                   static_cast<uint32_t>(*(payload_.first + 34)));
  return static_cast<double>(v);
}
double VescPacketValues::amp_hours_charged() const
{
  int32_t v = static_cast<int32_t>((static_cast<uint32_t>(*(payload_.first + 35)) << 24) +
                                   (static_cast<uint32_t>(*(payload_.first + 36)) << 16) +
                                   (static_cast<uint32_t>(*(payload_.first + 37)) << 8) +
                                   static_cast<uint32_t>(*(payload_.first + 38)));
  return static_cast<double>(v);
}
double VescPacketValues::watt_hours() const
{
  int32_t v = static_cast<int32_t>((static_cast<uint32_t>(*(payload_.first + 39)) << 24) +
                                   (static_cast<uint32_t>(*(payload_.first + 40)) << 16) +
                                   (static_cast<uint32_t>(*(payload_.first + 41)) << 8) +
                                   static_cast<uint32_t>(*(payload_.first + 42)));
  return static_cast<double>(v);
}
double VescPacketValues::watt_hours_charged() const
{
  int32_t v = static_cast<int32_t>((static_cast<uint32_t>(*(payload_.first + 43)) << 24) +
                                   (static_cast<uint32_t>(*(payload_.first + 44)) << 16) +
                                   (static_cast<uint32_t>(*(payload_.first + 45)) << 8) +
                                   static_cast<uint32_t>(*(payload_.first + 46)));
  return static_cast<double>(v);
}
double VescPacketValues::tachometer() const
{
  int32_t v = static_cast<int32_t>((static_cast<uint32_t>(*(payload_.first + 47)) << 24) +
                                   (static_cast<uint32_t>(*(payload_.first + 48)) << 16) +
                                   (static_cast<uint32_t>(*(payload_.first + 49)) << 8) +
                                   static_cast<uint32_t>(*(payload_.first + 50)));
  return static_cast<double>(v);
}
double VescPacketValues::tachometer_abs() const
{
  int32_t v = static_cast<int32_t>((static_cast<uint32_t>(*(payload_.first + 51)) << 24) +
                                   (static_cast<uint32_t>(*(payload_.first + 52)) << 16) +
                                   (static_cast<uint32_t>(*(payload_.first + 53)) << 8) +
                                   static_cast<uint32_t>(*(payload_.first + 54)));
  return static_cast<double>(v);
}
int VescPacketValues::fault_code() const
{
  return static_cast<int32_t>(*(payload_.first + 55));
}

REGISTER_PACKET_TYPE(COMM_GET_VALUES, VescPacketValues)

VescPacketRequestValues::VescPacketRequestValues() :
  VescPacket("RequestFWVersion", 1, COMM_GET_VALUES)
{
  VescFrame::CRC crc_calc;
  crc_calc.process_bytes(&(*payload_.first), boost::distance(payload_));
  uint16_t crc = crc_calc.checksum();
  *(frame_->end() - 3) = static_cast<uint8_t>(crc >> 8);
  *(frame_->end() - 2) = static_cast<uint8_t>(crc & 0xFF);
}

/*------------------------------------------------------------------------------------------------*/


VescPacketSetDuty::VescPacketSetDuty(double duty) :
  VescPacket("SetDuty", 5, COMM_SET_DUTY)
{
  /** @todo range check duty */

  int32_t v = static_cast<int32_t>(duty * 100000.0);

  *(payload_.first + 1) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 24) & 0xFF);
  *(payload_.first + 2) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 16) & 0xFF);
  *(payload_.first + 3) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 8) & 0xFF);
  *(payload_.first + 4) = static_cast<uint8_t>(static_cast<uint32_t>(v) & 0xFF);

  VescFrame::CRC crc_calc;
  crc_calc.process_bytes(&(*payload_.first), boost::distance(payload_));
  uint16_t crc = crc_calc.checksum();
  *(frame_->end() - 3) = static_cast<uint8_t>(crc >> 8);
  *(frame_->end() - 2) = static_cast<uint8_t>(crc & 0xFF);
}


VescPacketSetDuty::VescPacketSetDuty(double duty, int m2_address) :
  VescPacket("CanSetDuty", 7, COMM_FORWARD_CAN)
  {
  int32_t v = static_cast<int32_t>(duty * 100000.0);
  *(payload_.first + 1) = static_cast<uint8_t>(m2_address);
  *(payload_.first + 2) = static_cast<uint8_t>(COMM_SET_DUTY);
  *(payload_.first + 3) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 24) & 0xFF);
  *(payload_.first + 4) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 16) & 0xFF);
  *(payload_.first + 5) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 8) & 0xFF);
  *(payload_.first + 6) = static_cast<uint8_t>(static_cast<uint32_t>(v) & 0xFF);
  VescFrame::CRC crc_calc;
  crc_calc.process_bytes(&(*payload_.first), boost::distance(payload_));
  uint16_t crc = crc_calc.checksum();
  *(frame_->end() - 3) = static_cast<uint8_t>(crc >> 8);
  *(frame_->end() - 2) = static_cast<uint8_t>(crc & 0xFF);
}

/*------------------------------------------------------------------------------------------------*/

VescPacketSetCurrent::VescPacketSetCurrent(double current) :
  VescPacket("SetCurrent", 5, COMM_SET_CURRENT)
{
  int32_t v = static_cast<int32_t>(current * 1000.0);

  *(payload_.first + 1) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 24) & 0xFF);
  *(payload_.first + 2) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 16) & 0xFF);
  *(payload_.first + 3) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 8) & 0xFF);
  *(payload_.first + 4) = static_cast<uint8_t>(static_cast<uint32_t>(v) & 0xFF);

  VescFrame::CRC crc_calc;
  crc_calc.process_bytes(&(*payload_.first), boost::distance(payload_));
  uint16_t crc = crc_calc.checksum();
  *(frame_->end() - 3) = static_cast<uint8_t>(crc >> 8);
  *(frame_->end() - 2) = static_cast<uint8_t>(crc & 0xFF);
}

/*------------------------------------------------------------------------------------------------*/

VescPacketSetCurrentBrake::VescPacketSetCurrentBrake(double current_brake) :
  VescPacket("SetCurrentBrake", 5, COMM_SET_CURRENT_BRAKE)
{
  int32_t v = static_cast<int32_t>(current_brake * 1000.0);

  *(payload_.first + 1) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 24) & 0xFF);
  *(payload_.first + 2) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 16) & 0xFF);
  *(payload_.first + 3) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 8) & 0xFF);
  *(payload_.first + 4) = static_cast<uint8_t>(static_cast<uint32_t>(v) & 0xFF);

  VescFrame::CRC crc_calc;
  crc_calc.process_bytes(&(*payload_.first), boost::distance(payload_));
  uint16_t crc = crc_calc.checksum();
  *(frame_->end() - 3) = static_cast<uint8_t>(crc >> 8);
  *(frame_->end() - 2) = static_cast<uint8_t>(crc & 0xFF);
}

/*------------------------------------------------------------------------------------------------*/

VescPacketSetRPM::VescPacketSetRPM(double rpm) :
  VescPacket("SetRPM", 5, COMM_SET_RPM)
{
  int32_t v = static_cast<int32_t>(rpm);

  *(payload_.first + 1) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 24) & 0xFF);
  *(payload_.first + 2) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 16) & 0xFF);
  *(payload_.first + 3) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 8) & 0xFF);
  *(payload_.first + 4) = static_cast<uint8_t>(static_cast<uint32_t>(v) & 0xFF);

  VescFrame::CRC crc_calc;
  crc_calc.process_bytes(&(*payload_.first), boost::distance(payload_));
  uint16_t crc = crc_calc.checksum();
  *(frame_->end() - 3) = static_cast<uint8_t>(crc >> 8);
  *(frame_->end() - 2) = static_cast<uint8_t>(crc & 0xFF);
}

/*------------------------------------------------------------------------------------------------*/

VescPacketSetPos::VescPacketSetPos(double pos) :
  VescPacket("SetPos", 5, COMM_SET_POS)
{
  /** @todo range check pos */

  int32_t v = static_cast<int32_t>(pos * 1000000.0);

  *(payload_.first + 1) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 24) & 0xFF);
  *(payload_.first + 2) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 16) & 0xFF);
  *(payload_.first + 3) = static_cast<uint8_t>((static_cast<uint32_t>(v) >> 8) & 0xFF);
  *(payload_.first + 4) = static_cast<uint8_t>(static_cast<uint32_t>(v) & 0xFF);

  VescFrame::CRC crc_calc;
  crc_calc.process_bytes(&(*payload_.first), boost::distance(payload_));
  uint16_t crc = crc_calc.checksum();
  *(frame_->end() - 3) = static_cast<uint8_t>(crc >> 8);
  *(frame_->end() - 2) = static_cast<uint8_t>(crc & 0xFF);
}

/*------------------------------------------------------------------------------------------------*/

VescPacketSetServoPos::VescPacketSetServoPos(double servo_pos) :
  VescPacket("SetServoPos", 3, COMM_SET_SERVO_POS)
{
  /** @todo range check pos */

  int16_t v = static_cast<int16_t>(servo_pos * 1000.0);

  *(payload_.first + 1) = static_cast<uint8_t>((static_cast<uint16_t>(v) >> 8) & 0xFF);
  *(payload_.first + 2) = static_cast<uint8_t>(static_cast<uint16_t>(v) & 0xFF);

  VescFrame::CRC crc_calc;
  crc_calc.process_bytes(&(*payload_.first), boost::distance(payload_));
  uint16_t crc = crc_calc.checksum();
  *(frame_->end() - 3) = static_cast<uint8_t>(crc >> 8);
  *(frame_->end() - 2) = static_cast<uint8_t>(crc & 0xFF);
}

} // namespace vesc_driver
